src/core/metrics.cc

```diff
@@ -232,9 +232,20 @@ bool metric_id::operator==(
 // Unfortunately, metrics_impl can not be shared because it
 // need to be available before the first users (reactor) will call it
 
-shared_ptr<impl>  get_local_impl() {
-    static thread_local auto the_impl = ::seastar::make_shared<impl>();
-    return the_impl;
+shared_ptr<impl> get_local_impl(int handle) {
+    auto& impls = get_metric_implementations();
+
+    auto impl_iter = std::find_if(impls.begin(), impls.end(), [handle](const auto& crnt) {
+        return crnt->get_handle() == handle;
+    });
+
+    if (impl_iter == impls.end()) {
+        auto ptr = ::seastar::make_shared<impl>(handle);
+        impls.push_back(ptr);
+        return ptr;
+    } else {
+        return *impl_iter;
+    }
```

Comment:
I think it might be a little less intrusive to have `impls` as a `std::vector<ss::shared_ptr<ss::metrics::impl>>`, and the handle an index into it.

Reply:
I think it's kind of nice that the application doesn't have to register the metrics prior to usage, but I don't feel too strongly about it.